libkbfs/folder_branch_ops.go

```diff
@@ -1863,19 +1863,95 @@ func (fbo *folderBranchOps) isRevisionConflict(err error) bool {
 		isConflictFolderMapping
 }
 
+// doPutWithCancellationCheck is a wrapper for any MD put Ops (e.g. Put and
+// PutUnmerged). If context is canceled during the Op, it checks the latest MD
+// and finds out whether the cancellation was successful (meaning the put
+// didn't go through) or unsuccessful (meaning the put did go through). This
+// allows us to maintain a state consistent with what's returned to application
+// and not conflict with ourselves.
+//
+// If MD puts succeeds, we should try to avoid any cancellation afterwards. As
+// a result, if put succeeds, doPutWithCancellationCheck also returns a new ctx
+// which adds extra grace period to the ctx for cancellation.
+//
+// If bid == NullBranchID, doPutWithCancellationCheck does a regular Put.
+// Otherwise, it does a PutUnmerged.
+//
+// A case that we are trying to solve here is: when a O_EXCL create is
+// interrupted by SIGALRM (e.g. in git). This results in a canceled context and
+// EINTR is returned. However, the MD put could still have gone through. But
+// with EINTR, the application assumes the file was not created and retries.
+// The retry would fail because previous create was successful.
+func (fbo *folderBranchOps) doPutWithCancellationCheck(
+	ctx context.Context, lState *lockState, md *RootMetadata, bid BranchID) (
+	mdID MdID, err error) {
+
+	if err = EnterCriticalWithTimeout(ctx, fbo.config.GracePeriod()); err != nil {
+		return mdID, err
+	}
+
+	oldRevision := fbo.getCurrMDRevision(lState)
+	var errPut error
+	if bid == NullBranchID {
+		errPut = fbo.config.MDOps().Put(ctx, md)
+	} else {
+		errPut = fbo.config.MDOps().PutUnmerged(ctx, md, bid)
+	}
+	mdID, err = fbo.config.Crypto().MakeMdID(&md.BareRootMetadata)
+	if err != nil {
+		return mdID, err
+	}
+	if errPut == context.Canceled {
```

Comment:
It's "critical aware" so it can only be `context.Canceled` right? But yet, no harm to check `context.DeadlineExceeded` as well